Mathlib/Analysis/SpecialFunctions/BinEntropy.lean

```diff
@@ -0,0 +1,508 @@
+noncomputable def binaryEntropy := qaryEntropy 2
+
+lemma binaryEntropy_eq : binaryEntropy = (fun p => -p * log p - (1 - p) * log (1 - p)) := by
+  have : (2 : ℝ) - 1 = 1 := by norm_num
+  ext
+  simp [binaryEntropy, qaryEntropy, this]
+
+lemma binaryEntropy_eq' {p : ℝ} : binaryEntropy p = -p * log p - (1 - p) * log (1 - p) := by
+  rw [binaryEntropy_eq]
+
+@[simp] lemma qaryEntropy_zero {q : ℕ} : qaryEntropy q 0 = 0 := by
+  simp only [qaryEntropy, zero_mul, log_zero, mul_zero, sub_self, sub_zero, log_one]
+
+@[simp] lemma binaryEntropy_one : binaryEntropy 1 = 0 := by
+  simp only [binaryEntropy_eq', log_one, mul_zero, sub_self, log_zero]
+
+@[simp] lemma qaryEntropy_one {q : ℕ} : qaryEntropy q 1 = log (q - 1) := by
+  unfold qaryEntropy
+  simp only [log_one, mul_zero, sub_self, log_zero, one_mul, sub_zero]
+
+@[simp] lemma binaryEntropy_onehalf : binaryEntropy 2⁻¹ = log 2 := by
+  simp only [binaryEntropy_eq', show (1 : ℝ) - 2⁻¹ = 2⁻¹ by norm_num, log_inv]
+  field_simp
+
+/-- `binaryEntropy` is symmetric about 1/2. -/
+@[simp] lemma binaryEntropy_one_sub (p : ℝ) :
+    binaryEntropy (1 - p) = binaryEntropy p := by
+  simp only [binaryEntropy_eq', neg_sub, sub_sub_cancel, neg_mul]
+  ring
+
+/-- `binaryEntropy` is symmetric about 1/2. -/
+lemma binaryEntropy_two_inv_add (p : ℝ) :
+    binaryEntropy (2⁻¹ + p) = binaryEntropy (2⁻¹ - p) := by
+  simp only [binaryEntropy_eq', neg_sub, sub_sub_cancel, neg_mul]
+  ring_nf
+
+lemma qaryEntropy_eq_log_mul_add_binaryEntropy {q : ℕ} {p : ℝ} :
+    qaryEntropy q p = p * log (q - 1) + binaryEntropy p := by
+  unfold binaryEntropy qaryEntropy
+  simp only [show ((2:ℕ) - (1:ℝ)).log = 0 by norm_num]
+  ring
+
+lemma qaryEntropy_eq_log_mul_add_binaryEntropy' {q : ℕ} :
+    qaryEntropy q = (fun p ↦ p * log (q - 1)) + binaryEntropy := by
+  ext
+  simp only [Pi.add_apply, qaryEntropy_eq_log_mul_add_binaryEntropy]
+
+lemma binaryEntropy_pos {p : ℝ} (pgt0 : 0 < p) (plt1 : p < 1) : 0 < binaryEntropy p := by
+  simp only [binaryEntropy_eq']
+  have pos_sum_pos_pos (a b : ℝ) (ha : 0 ≤ a) (hb : b < 0) : 0 < a - b := by linarith
+  refine pos_sum_pos_pos (-p * log p) ((1 - p) * log (1 - p)) ?_ ?_
+  · rw [show -p * log p = p * (-log p) by ring]
+    exact (Real.mul_pos (by linarith) (by linarith [log_neg pgt0 plt1])).le
+  · exact mul_neg_of_pos_of_neg (by linarith) (log_neg (by linarith) (by linarith))
+
+lemma qaryEntropy_pos {q : ℕ} {p : ℝ} (pgt0 : 0 < p) (plt1 : p < 1) : 0 < qaryEntropy q p := by
+  unfold qaryEntropy
+  have p_q_log_nonneg : 0 ≤ p * ((q : ℝ) - 1).log := by
+    rw [mul_nonneg_iff_of_pos_left pgt0, show q - (1 : ℝ) = (q - 1 : ℤ) by norm_cast]
+    exact Real.log_intCast_nonneg _
+  have rest_is_pos : 0 < -(p * p.log) - (1 - p) * (1 - p).log := by
+    simp only [← neg_mul, ← binaryEntropy_eq']
+    exact binaryEntropy_pos pgt0 plt1
+  linarith
+
+/- Outside usual range of `binaryEntropy`. This is due to `log x = log |x|` -/
```

Comment:
Did you mean for this to be a docstring?

Reply:
Yes. I'm sure it could be improved. As a beginner, I tend to think it important to point out in docstrings how "junk values" behave and where things work differently than what you would see in normal maths.

Reply:
I think the point is that if it's a doctring, it should start with two dashes.

Reply:
Thanks, done.